# Simple Market Basket Analysis
Market Basket Analysis es una técnica que nos permite descubrir reglas de asociación para productos relacionados, en un entorno de mercado. En este tipo de entornos cada compra representa una única compra de una serie d eproductos y deseamos encontrar cual es la relación causal entre los diferentes productos.

## Ejercicio 1
Suponga que hay 100 artículos, numerados del 1 al 100; también 100 canastas, igualmente numeradas del 1 al 100. El artículo i está en la canasta b, si y sólo si i divide b sin ningún resto. Por lo tanto, el artículo 1 está en todas las canastas, el artículo 2 está en las cincuenta canastas pares, y así sucesivamente. La cesta 12 consta de elementos {1, 2, 3, 4, 6, 12}, ya que estos son todos los enteros que dividen 12. Responda las siguientes preguntas:

### Ejercicio 1a
Si el umbral de soporte es 5, ¿qué elementos son frecuentes?

Crearemos una función para generar una canasta determinada, viendo que son todos los números de los 100 artículos que dividen de manera exacta a la canasta i

In [2]:
def generaCanasta(i):
  productos = []
  for producto in range(1, 101):
    if i % producto == 0:
      productos.append(producto)
  return productos

print(generaCanasta(12))

[1, 2, 3, 4, 6, 12]


Podemos generar todas las canastas utilizando un ciclo iterado simple

In [8]:
canastas_ejercicio_1 = []
for i in range(1, 101):
    canastas_ejercicio_1.append(generaCanasta(i))

for i, c in enumerate(canastas_ejercicio_1):
    print("Canasta " + str(i + 1) + ": " + str(c))

Canasta 1: [1]
Canasta 2: [1, 2]
Canasta 3: [1, 3]
Canasta 4: [1, 2, 4]
Canasta 5: [1, 5]
Canasta 6: [1, 2, 3, 6]
Canasta 7: [1, 7]
Canasta 8: [1, 2, 4, 8]
Canasta 9: [1, 3, 9]
Canasta 10: [1, 2, 5, 10]
Canasta 11: [1, 11]
Canasta 12: [1, 2, 3, 4, 6, 12]
Canasta 13: [1, 13]
Canasta 14: [1, 2, 7, 14]
Canasta 15: [1, 3, 5, 15]
Canasta 16: [1, 2, 4, 8, 16]
Canasta 17: [1, 17]
Canasta 18: [1, 2, 3, 6, 9, 18]
Canasta 19: [1, 19]
Canasta 20: [1, 2, 4, 5, 10, 20]
Canasta 21: [1, 3, 7, 21]
Canasta 22: [1, 2, 11, 22]
Canasta 23: [1, 23]
Canasta 24: [1, 2, 3, 4, 6, 8, 12, 24]
Canasta 25: [1, 5, 25]
Canasta 26: [1, 2, 13, 26]
Canasta 27: [1, 3, 9, 27]
Canasta 28: [1, 2, 4, 7, 14, 28]
Canasta 29: [1, 29]
Canasta 30: [1, 2, 3, 5, 6, 10, 15, 30]
Canasta 31: [1, 31]
Canasta 32: [1, 2, 4, 8, 16, 32]
Canasta 33: [1, 3, 11, 33]
Canasta 34: [1, 2, 17, 34]
Canasta 35: [1, 5, 7, 35]
Canasta 36: [1, 2, 3, 4, 6, 9, 12, 18, 36]
Canasta 37: [1, 37]
Canasta 38: [1, 2, 19, 38]
Canasta 39: [1, 3, 13, 39]
Canasta 

Podemos generar el soporte de los artículos contando en cuantas canastas aparece. El producto 1 aparece en las 100 canastas, el artículo 2 en las canastas pares, y así sucesivamente. En general el soporte para el artículo i es la parte entera de la división de $\frac{100}{i}$

In [9]:
soportes = []
for i in range(1, 101):
    soportes.append(100//i)

for i, soporte in enumerate(soportes):
  print("Soporte articulo " + str(i + 1) + ":" + str(soporte))

Soporte articulo 1:100
Soporte articulo 2:50
Soporte articulo 3:33
Soporte articulo 4:25
Soporte articulo 5:20
Soporte articulo 6:16
Soporte articulo 7:14
Soporte articulo 8:12
Soporte articulo 9:11
Soporte articulo 10:10
Soporte articulo 11:9
Soporte articulo 12:8
Soporte articulo 13:7
Soporte articulo 14:7
Soporte articulo 15:6
Soporte articulo 16:6
Soporte articulo 17:5
Soporte articulo 18:5
Soporte articulo 19:5
Soporte articulo 20:5
Soporte articulo 21:4
Soporte articulo 22:4
Soporte articulo 23:4
Soporte articulo 24:4
Soporte articulo 25:4
Soporte articulo 26:3
Soporte articulo 27:3
Soporte articulo 28:3
Soporte articulo 29:3
Soporte articulo 30:3
Soporte articulo 31:3
Soporte articulo 32:3
Soporte articulo 33:3
Soporte articulo 34:2
Soporte articulo 35:2
Soporte articulo 36:2
Soporte articulo 37:2
Soporte articulo 38:2
Soporte articulo 39:2
Soporte articulo 40:2
Soporte articulo 41:2
Soporte articulo 42:2
Soporte articulo 43:2
Soporte articulo 44:2
Soporte articulo 45:2
Soporte 

Seleccionamos aquellos artículos con un soporte de 5 o mayor

In [12]:
items_frecuentes = []
for i, soporte in enumerate(soportes):
    if soporte >= 5:
        items_frecuentes.append(i+1)
print(items_frecuentes)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


Como vemos los primeros 20 items son los que aparecen 5 veces o más en las cestas

### Ejercicio 1b
¿Cuál es la suma de los tamaños de todas las canastas?

Si sumamos todos los soportes podemos obtener el tamaño de todas las canastas

In [15]:
import numpy as np
print(np.sum(soportes))

482


Como vemos hay 482 artículos

## Ejercicio 2
Suponga que hay 100 artículos, numerados del 1 al 100; también 100 canastas, igualmente numeradas del 1 al 100. El artículo i está en la canasta b, si y sólo si b divide i sin resto. Por ejemplo, la canasta 12 consta de artículos {12, 24, 36, 48, 60, 72, 84, 96}.

## Ejercicio 2a
En este caso la canasta está compuesta por todos los múltiplos del número de canasta que sean menores o iguales a 100. Para generar una canasta solo tenemos que generar los múltiplos menores a 100

In [20]:
def generaCanasta2(i):
  productos = []
  for producto in range(1, 100//i + 1):
        productos.append(i*producto)
  return productos

print(generaCanasta2(12))

[12, 24, 36, 48, 60, 72, 84, 96]


Generemos todas las canastas para el ejercicio 2

In [22]:
canastas_ejercicio_2 = []
for i in range(1, 101):
  canastas_ejercicio_2.append(generaCanasta2(i))

for i, canasta in enumerate(canastas_ejercicio_2):
    print("Canasta " + str(i + 1) + ": " + str(canasta))

Canasta 1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]
Canasta 2: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100]
Canasta 3: [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99]
Canasta 4: [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100]
Canasta 5: [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100]
Canasta 6: [6, 12, 18, 24, 30, 36, 42, 48, 

Podemos calcular el tamaño de las canastas sumando el tamaño de cada una de ellas:

In [23]:
np.sum([len(canasta) for canasta in canastas_ejercicio_2])

482

Como vemos también suman 482 artículos en total

### Verificar si varios productos están en una canasta
Escribamos el código para verificar en que canastas están juntos los productos que pasemos

In [30]:
def inCanasta(numeros, canastas):
  canastas_aparece = []
  for c, canasta in enumerate(canastas):
    append = True
    for n in numeros:
        if n not in canasta:
            append = False
            break
    if append:
        canastas_aparece.append(c + 1)
  return canastas_aparece

print("Canastas en donde aparece el 24")
print(inCanasta([24], canastas_ejercicio_2))

print("Canastas en donde aparece el 60")
print(inCanasta([60], canastas_ejercicio_2))

print("Canastas en donde aparecen tanto el 24 como el 60")
print(inCanasta([24, 60], canastas_ejercicio_2))

Canastas en donde aparece el 24
[1, 2, 3, 4, 6, 8, 12, 24]
Canastas en donde aparece el 60
[1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30, 60]
Canastas en donde aparecen tanto el 24 como el 60
[1, 2, 3, 4, 6, 12]


## Ejercicio 3
Para los datos del Ejercicio 1, ¿cuál es la confianza de la siguiente regla de asociación?
$\{5,7\} \rightarrow 2$

Para resolver este ejercicio debemos tener en cuenta que la fórmula para calcular la confianza es:

$$ confianza(A \rightarrow B) = \frac{soporte(A \rightarrow B)}{soporte(B}$$

Recordemos que el soporte de $A \rightarrow B$ es el número de veces que A y B aparecen juntos, en este caso para calcular el soporte de $\{5,7\}$ tenemos que contar las canastas en donde el producto 5 y el producto 7 aparecen juntos. Así mismo para calcular el soporte de $\{5,7\} \rightarrow 2$, debemos contar el número de veces que 5, 7 y 2 aparecen juntos:


In [31]:
print("Canastas en donde aparecen el 5 y el 7")
print(inCanasta([5,7], canastas_ejercicio_1))

print("Canastas en donde aparece el 5 el 7 y el 2")
print(inCanasta([5,7,2], canastas_ejercicio_1))

Canastas en donde aparecen el 5 y el 7
[35, 70]
Canastas en donde aparece el 5 el 7 y el 2
[70]


Calculamos el soporte

In [35]:
soporte = lambda numeros, canastas: len(inCanasta(numeros, canastas))
support_5_7 = soporte([5,7], canastas_ejercicio_1)
support_5_7_2 = soporte([5,7,2], canastas_ejercicio_1)
print("Soporte de 5 y 7: " + str(support_5_7))
print("Soporte de 5, 7 --> 2: " + str(support_5_7_2))

Soporte de 5 y 7: 2
Soporte de 5, 7 --> 2: 1


Calculamos la confianza

In [37]:
support_5_7_2/support_5_7

0.5

## Ejercicio 4
Para los datos del Ejercicio 2, ¿cuál es la confianza de las siguientes reglas de asociación?

### Ejercicio 4a
$\{24,60\} \rightarrow 8$

Siguiendo un proceso análogo al presentado arriba podemos ver las canastas y calcular los soportes de la sigueinte manera

In [38]:
print("Canastas en donde aparecen el 24 y el 60")
print(inCanasta([24,60], canastas_ejercicio_2))

print("Canastas en donde aparece el 24 el 60 y el 8")
print(inCanasta([24,60,8], canastas_ejercicio_2))

support_24_60 = soporte([24,60], canastas_ejercicio_2)
support_24_60_8 = soporte([24,60,8], canastas_ejercicio_2)
print("Soporte de {24, 60}")
print(support_24_60)

print("Soporte de {24, 60} --> 8")
print(support_24_60_8)

print("Confianza {24, 60} --> 8")
print(support_24_60_8/support_24_60)

Canastas en donde aparecen el 24 y el 60
[1, 2, 3, 4, 6, 12]
Canastas en donde aparece el 24 el 60 y el 8
[1, 2, 4]
Soporte de {24, 60}
6
Soporte de {24, 60} --> 8
3
Confianza {24, 60} --> 8
0.5


### Ejercicio 4b
$\{2, 3, 4\} \rightarrow 5$

Realizamos el mismo proceso que hicimos anteriormente

In [39]:
print("Canastas en donde aparecen el 2, 3 y 4")
print(inCanasta([2, 3, 4], canastas_ejercicio_2))

print("Canastas en donde aparece el 2, 3, 4 y 5")
print(inCanasta([2,3,4,5], canastas_ejercicio_2))

support_2_3_4 = soporte([2,3,4], canastas_ejercicio_2)
support_2_3_4_5 = soporte([2,3,4,5], canastas_ejercicio_2)
print("Soporte de {2,3,4}")
print(support_2_3_4)

print("Soporte de {2,3,4} --> 5")
print(support_2_3_4_5)

print("Confianza {2,3, 4} --> 5")
print(support_2_3_4_5/support_2_3_4)

Canastas en donde aparecen el 2, 3 y 4
[1]
Canastas en donde aparece el 2, 3, 4 y 5
[1]
Soporte de {2,3,4}
1
Soporte de {2,3,4} --> 5
1
Confianza {2,3, 4} --> 5
1.0
